predicators/spot_utils/perception_utils.py

```diff
@@ -0,0 +1,444 @@
+    """
+    plt.figure(figsize=(10, 10))
+    plt.imshow(im)
+    for mask in masks:
+        show_mask(mask, plt.gca(), random_color=True)
+    for box, class_name, score in zip(input_boxes, classes, scores):
+        show_box(box, plt.gca())
+        x, y = box[:2]
+        plt.gca().text(x,
+                       y - 5,
+                       class_name + f': {score:.2f}',
+                       color='white',
+                       fontsize=12,
+                       fontweight='bold',
+                       bbox=dict(facecolor='green',
+                                 edgecolor='green',
+                                 alpha=0.5))
+    plt.axis('off')
+    plt.show()
+
+
+def show_mask(mask: np.ndarray,
+              ax: matplotlib.axes.Axes,
+              random_color: bool = False) -> None:
+    """Helper function for visualization that displays a segmentation mask."""
+    if random_color:
+        color = np.concatenate([np.random.random(3), np.array([0.6])], axis=0)
+    else:
+        color = np.array([30 / 255, 144 / 255, 255 / 255, 0.6])
+    h, w = mask.shape[-2:]
+    mask_image = mask.reshape((h, w, 1)) * color.reshape((1, 1, -1))
+    ax.imshow(mask_image)
+
+
+def show_box(box: np.ndarray, ax: matplotlib.axes.Axes) -> None:
+    """Helper function for visualization that displays a bounding box."""
+    x0, y0 = box[0], box[1]
+    w, h = box[2] - box[0], box[3] - box[1]
+    ax.add_patch(
+        plt.Rectangle((x0, y0),
+                      w,
+                      h,
+                      edgecolor='green',
+                      facecolor=(0, 0, 0, 0),
+                      lw=2))
+
+
+def query_sam(image_in: np.ndarray, classes: List[str],
+              viz: bool) -> Optional[Dict[str, List[np.ndarray]]]:
+    """Send a query to SAM and return the response.
+
+    The response is a dictionary that contains 4 keys: 'boxes',
+    'classes', 'masks' and 'scores'.
+    """
+    image = Image.fromarray(image_in)
+    buf = image_to_bytes(image)
+    r = requests.post("http://localhost:5550/predict",
+                      files={"file": buf},
+                      data={"classes": ",".join(classes)})
+
+    # If the status code is not 200, then fail.
+    if r.status_code != 200:
+        return None
+
+    with io.BytesIO(r.content) as f:
+        arr = np.load(f, allow_pickle=True)
+        boxes = arr['boxes']
+        ret_classes = arr['classes']
+        masks = arr['masks']
+        scores = arr['scores']
+
+    d = {
+        "boxes": boxes,
+        "classes": ret_classes,
+        "masks": masks,
+        "scores": scores
+    }
+
+    if viz:
+        # Optional visualization useful for debugging.
+        visualize_output(image, d["masks"], d["boxes"], d["classes"],
+                         d["scores"])
+
+    # Filter out detections by confidence. We threshold detections
+    # at a 40% confidence level minimum, and if there are multiple
+    #, we only select the most confident one.
+    k = 1
```

Comment:
can we simplify this logic for now? we can change it back later if we want to use k>1 for some reason.

Reply:
i'm actually not that sure how to simplify this; it's a somewhat complex operation to threshold and then get the index that corresponds to the max value. idk how to do that in pure numpy without for-looping, but maybe I'm missing something

Reply:
nvm - resolved offline